File: core/src/nl/t64/game/rpg/screens/inventory/tooltip/ItemSlotTooltip.java
package nl.t64.game.rpg.screens.inventory.tooltip;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.NinePatch;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.scenes.scene2d.actions.Actions;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.utils.NinePatchDrawable;
import com.badlogic.gdx.utils.Align;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.components.party.*;
import nl.t64.game.rpg.constants.Constant;
import nl.t64.game.rpg.screens.inventory.InventoryUtils;
import nl.t64.game.rpg.screens.inventory.itemslot.InventoryImage;
import nl.t64.game.rpg.screens.inventory.itemslot.ItemSlot;

import java.util.List;
import java.util.Optional;


public class ItemSlotTooltip extends BaseTooltip {

    private static final float SLOT_SIZE = 64f;
    private static final float THREE_QUARTERS = SLOT_SIZE * 0.75f;
    private static final String RIGHT_BORDER = "sprites/tooltip_right.png";
    private static final float COLUMN_SPACING = 20f;
    private static final float HALF_SPACING = 10f;
    private static final String EMPTY_ROW = "";
    private static final String LEFT_TITLE = EMPTY_ROW;
    private static final String RIGHT_TITLE = "Currently Equipped";
    private static final Color ORANGE = new Color(0xFF9000FF);
    private static final float DELAY = 0.5f;

    @Override
    public void toggle(ItemSlot itemSlot) {
        if (itemSlot.hasItem()) {
            boolean isEnabled = Utils.getGameData().isTooltipEnabled();
            Utils.getGameData().setTooltipEnabled(!isEnabled);
            setupTooltip(itemSlot);
            window.setVisible(!isEnabled);
        }
    }

    @Override
    public void toggleCompare(ItemSlot itemSlot) {
        if (itemSlot.hasItem() && !window.hasActions()) {
            boolean isEnabled = Utils.getGameData().isComparingEnabled();
            Utils.getGameData().setComparingEnabled(!isEnabled);
            updateDescription(itemSlot);
        }
    }

    public void refresh(ItemSlot itemSlot) {
        hide();
        if (Utils.getGameData().isTooltipEnabled() && itemSlot.hasItem()) {
            setupTooltip(itemSlot);
            window.addAction(Actions.sequence(Actions.delay(DELAY),
                                              Actions.show()));
        }
    }

    private void setupTooltip(ItemSlot itemSlot) {
        var localCoords = new Vector2(itemSlot.getOriginX(), itemSlot.getOriginY());
        itemSlot.localToStageCoordinates(localCoords);
        updateDescription(itemSlot);
        window.setPosition(localCoords.x + THREE_QUARTERS, localCoords.y + THREE_QUARTERS);
        window.toFront();
    }

    void updateDescription(ItemSlot itemSlot) {
        window.clear();
        if (itemSlot.hasItem()) {
            final InventoryImage hoveredImage = itemSlot.getCertainInventoryImage();
            final InventoryItem hoveredItem = hoveredImage.inventoryItem;
            final InventoryGroup inventoryGroup = hoveredItem.getGroup();
            final HeroItem selectedHero = InventoryUtils.getSelectedHero();
            final Optional<String> errorMessage = selectedHero.createMessageIfHeroHasNotEnoughFor(hoveredItem);
            final Optional<InventoryItem> equippedItem = selectedHero.getInventoryItem(inventoryGroup);

            if (itemSlot.isOnHero()) {
                createSingleTooltip(hoveredImage);
            } else if (inventoryGroup.equals(InventoryGroup.RESOURCE)
                       || inventoryGroup.equals(InventoryGroup.POTION)
                       || inventoryGroup.equals(InventoryGroup.ITEM)) {
                createResourceTooltip(hoveredImage);
            } else if (errorMessage.isEmpty()
                       && equippedItem.isPresent()
                       && Utils.getGameData().isComparingEnabled()) {
                createDualTooltip(hoveredImage, new InventoryImage(equippedItem.get()));
            } else {
                createSingleTooltip(hoveredImage);
            }
        }
        window.pack();
    }

    void createResourceTooltip(InventoryImage inventoryImage) {
        Table hoveredTable = createDefaultTooltip(inventoryImage);
        window.add(hoveredTable);

        window.add().row();
        window.add(createLabel(EMPTY_ROW, Color.WHITE)).row();
        String description = String.join(System.lineSeparator(), inventoryImage.inventoryItem.getDescription());
        window.add(createLabel(description, Color.WHITE));
    }

    void createSingleTooltip(InventoryImage inventoryImage) {
        Table hoveredTable = createDefaultTooltip(inventoryImage);
        addPossibleDescription(inventoryImage, hoveredTable);
        window.add(hoveredTable);
    }

    private Table createDefaultTooltip(InventoryImage inventoryImage) {
        final var hoveredTable = new Table();
        hoveredTable.defaults().align(Align.left);

        final int totalMerchant = Utils.getGameData().getParty().getSumOfSkill(SkillItemId.MERCHANT);
        List<InventoryDescription> descriptionList = inventoryImage.getSingleDescription(totalMerchant);
        removeLeftUnnecessaryAttributes(descriptionList);
        descriptionList.forEach(attribute -> addToTable(hoveredTable, attribute, createSingleLabelStyle(attribute)));
        return hoveredTable;
    }

    private void createDualTooltip(InventoryImage hoveredImage, InventoryImage equippedImage) {
        final Table hoveredTable = createLeftTooltip(hoveredImage, equippedImage);
        final Table equippedTable = createRightTooltip(equippedImage, hoveredImage);
        window.add(hoveredTable);
        window.add().spaceRight(HALF_SPACING);
        window.add(equippedTable);
    }

    private Table createLeftTooltip(InventoryImage hoveredImage, InventoryImage equippedImage) {
        final var hoveredTable = new Table(window.getSkin());
        setRightBorder(hoveredTable);
        hoveredTable.padRight(HALF_SPACING);
        hoveredTable.defaults().align(Align.left);
        hoveredTable.add(createLabel(LEFT_TITLE, Color.WHITE)).row();

        final int totalMerchant = Utils.getGameData().getParty().getSumOfSkill(SkillItemId.MERCHANT);
        List<InventoryDescription> descriptionList = hoveredImage.getDualDescription(equippedImage, totalMerchant);
        removeLeftUnnecessaryAttributes(descriptionList);
        descriptionList.forEach(attribute -> addToTable(hoveredTable, attribute, createLeftLabelStyle(attribute)));
        addPossibleDescription(hoveredImage, hoveredTable);
        return hoveredTable;
    }

    private Table createRightTooltip(InventoryImage equippedImage, InventoryImage hoveredImage) {
        final var equippedTable = new Table();
        equippedTable.defaults().align(Align.left);
        equippedTable.add(createLabel(RIGHT_TITLE, Color.LIGHT_GRAY)).row();

        final int totalMerchant = Utils.getGameData().getParty().getSumOfSkill(SkillItemId.MERCHANT);
        List<InventoryDescription> descriptionList = equippedImage.getDualDescription(hoveredImage, totalMerchant);
        removeRightUnnecessaryAttributes(descriptionList);
        descriptionList.forEach(attribute -> addToTable(equippedTable, attribute, createRightLabelStyle(attribute)));
        addPossibleDescription(equippedImage, equippedTable);
        return equippedTable;
    }

    void addPossibleDescription(InventoryImage inventoryImage, Table table) {
        if (inventoryImage.inventoryItem.getDescription() != null) {
            String description = String.join(System.lineSeparator(), inventoryImage.inventoryItem.getDescription());
            table.add(createLabel(description, Color.VIOLET)).colspan(2);
        }
    }

    void addToTable(Table hoveredTable, InventoryDescription attribute, Label.LabelStyle labelStyle) {
        hoveredTable.add(new Label(getKey(attribute), labelStyle)).spaceRight(COLUMN_SPACING);
        hoveredTable.add(new Label(getValue(attribute), labelStyle)).row();
    }

    void removeLeftUnnecessaryAttributes(List<InventoryDescription> descriptionList) {
        removeBuy(descriptionList);
        removeSell(descriptionList);
    }

    void removeRightUnnecessaryAttributes(List<InventoryDescription> descriptionList) {
        removeBuy(descriptionList);
        removeSell(descriptionList);
    }

    void removeBuy(List<InventoryDescription> descriptionList) {
        descriptionList.removeIf(attribute -> attribute.getKey().equals(Constant.DESCRIPTION_KEY_BUY));
        descriptionList.removeIf(attribute -> attribute.getKey().equals(Constant.DESCRIPTION_KEY_BUY_PIECE));
        descriptionList.removeIf(attribute -> attribute.getKey().equals(Constant.DESCRIPTION_KEY_BUY_TOTAL));
    }

    void removeSell(List<InventoryDescription> descriptionList) {
        descriptionList.removeIf(attribute -> attribute.getKey().equals(Constant.DESCRIPTION_KEY_SELL));
        descriptionList.removeIf(attribute -> attribute.getKey().equals(Constant.DESCRIPTION_KEY_SELL_PIECE));
        descriptionList.removeIf(attribute -> attribute.getKey().equals(Constant.DESCRIPTION_KEY_SELL_TOTAL));
    }

    Label.LabelStyle createSingleLabelStyle(InventoryDescription attribute) {
        if (isBuyOrSellValue(attribute)) {
            return createLabelStyle(Color.GOLD);
        }
        return switch (attribute.getCompare()) {
            case SAME, MORE -> createLabelStyle(Color.WHITE);
            case LESS -> createLabelStyle(Color.RED);
        };
    }

    private Label.LabelStyle createLeftLabelStyle(InventoryDescription attribute) {
        if (isBuyOrSellValue(attribute)) {
            return createLabelStyle(Color.GOLD);
        }
        return switch (attribute.getCompare()) {
            case SAME -> createLabelStyle(Color.WHITE);
            case LESS -> createLabelStyle(ORANGE);
            case MORE -> createLabelStyle(Color.LIME);
        };
    }

    private Label.LabelStyle createRightLabelStyle(InventoryDescription attribute) {
        if (isBuyOrSellValue(attribute)) {
            return createLabelStyle(Color.GOLD);
        }
        return createLabelStyle(Color.WHITE);
    }

    private String getKey(InventoryDescription attribute) {
        if (attribute.getKey() instanceof SuperEnum attributeKey) {
            return attributeKey.getTitle();
        } else {
            return String.valueOf(attribute.getKey());
        }
    }

    private String getValue(InventoryDescription attribute) {
        if (attribute.getValue() instanceof SkillItemId attributeValue) {
            return attributeValue.getTitle();
        } else if (attribute.getKey().equals(CalcAttributeId.BASE_HIT)) {
            return String.format("%s%%", attribute.getValue());
        } else {
            return String.valueOf(attribute.getValue());
        }
    }

    private void setRightBorder(Table table) {
        final var texture = Utils.getResourceManager().getTextureAsset(RIGHT_BORDER);
        final var ninepatch = new NinePatch(texture, 0, 1, 0, 0);
        final var drawable = new NinePatchDrawable(ninepatch);
        table.setBackground(drawable);
    }

    private Label createLabel(String text, Color color) {
        return new Label(text, createLabelStyle(color));
    }

    private Label.LabelStyle createLabelStyle(Color color) {
        return new Label.LabelStyle(new BitmapFont(), color);
    }

    private boolean isBuyOrSellValue(InventoryDescription attribute) {
        return attribute.getKey().equals(Constant.DESCRIPTION_KEY_BUY_TOTAL)
               || attribute.getKey().equals(Constant.DESCRIPTION_KEY_SELL_TOTAL)
               || attribute.getKey().equals(Constant.DESCRIPTION_KEY_BUY_PIECE)
               || attribute.getKey().equals(Constant.DESCRIPTION_KEY_SELL_PIECE)
               || attribute.getKey().equals(Constant.DESCRIPTION_KEY_BUY)
               || attribute.getKey().equals(Constant.DESCRIPTION_KEY_SELL);
    }

}

File: core/src/nl/t64/game/rpg/screens/inventory/tooltip/ItemSlotTooltip.kt
package nl.t64.game.rpg.screens.inventory.tooltip

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.scenes.scene2d.actions.Actions
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.utils.Align
import nl.t64.game.rpg.Utils
import nl.t64.game.rpg.Utils.gameData
import nl.t64.game.rpg.components.party.CalcAttributeId
import nl.t64.game.rpg.components.party.SuperEnum
import nl.t64.game.rpg.components.party.inventory.InventoryDescription
import nl.t64.game.rpg.components.party.inventory.InventoryGroup
import nl.t64.game.rpg.components.party.inventory.ThreeState
import nl.t64.game.rpg.components.party.skills.SkillItemId
import nl.t64.game.rpg.constants.Constant
import nl.t64.game.rpg.screens.inventory.InventoryUtils
import nl.t64.game.rpg.screens.inventory.itemslot.InventoryImage
import nl.t64.game.rpg.screens.inventory.itemslot.ItemSlot


private const val SLOT_SIZE = 64f
private const val THREE_QUARTERS = SLOT_SIZE * 0.75f
private const val COLUMN_SPACING = 20f
private const val HALF_SPACING = 10f
private const val EMPTY_ROW = ""
private const val LEFT_TITLE = EMPTY_ROW
private const val RIGHT_TITLE = "Currently Equipped"
private val ORANGE = Color(-0x6fff01)
private const val DELAY = 0.5f

open class ItemSlotTooltip : BaseTooltip() {

    override fun toggle(itemSlot: ItemSlot?) {
        if (itemSlot?.hasItem() == true) {
            val isEnabled = gameData.isTooltipEnabled
            gameData.isTooltipEnabled = !isEnabled
            setupTooltip(itemSlot)
            window.isVisible = !isEnabled
        }
    }

    override fun toggleCompare(itemSlot: ItemSlot?) {
        if (itemSlot?.hasItem() == true && !window.hasActions()) {
            val isEnabled = gameData.isComparingEnabled
            gameData.isComparingEnabled = !isEnabled
            updateDescription(itemSlot)
        }
    }

    fun refresh(itemSlot: ItemSlot) {
        hide()
        if (gameData.isTooltipEnabled && itemSlot.hasItem()) {
            setupTooltip(itemSlot)
            window.addAction(Actions.sequence(Actions.delay(DELAY),
                                              Actions.show()))
        }
    }

    private fun setupTooltip(itemSlot: ItemSlot) {
        val localCoords = Vector2(itemSlot.originX, itemSlot.originY)
        itemSlot.localToStageCoordinates(localCoords)
        updateDescription(itemSlot)
        window.setPosition(localCoords.x + THREE_QUARTERS, localCoords.y + THREE_QUARTERS)
        window.toFront()
    }

    open fun updateDescription(itemSlot: ItemSlot) {
        window.clear()
        if (itemSlot.hasItem()) {
            val hoveredImage = itemSlot.getCertainInventoryImage()
            val hoveredItem = hoveredImage.inventoryItem
            val inventoryGroup = hoveredItem.group
            val selectedHero = InventoryUtils.getSelectedHero()
            val errorMessage = selectedHero.createMessageIfHeroHasNotEnoughFor(hoveredItem)
            val equippedItem = selectedHero.getInventoryItem(inventoryGroup)

            if (itemSlot.isOnHero()) {
                createSingleTooltip(hoveredImage)
            } else if (inventoryGroup == InventoryGroup.RESOURCE
                || inventoryGroup == InventoryGroup.POTION
                || inventoryGroup == InventoryGroup.ITEM
            ) {
                createResourceTooltip(hoveredImage)
            } else if (errorMessage == null
                && equippedItem != null
                && gameData.isComparingEnabled
            ) {
                createDualTooltip(hoveredImage, InventoryImage(equippedItem))
            } else {
                createSingleTooltip(hoveredImage)
            }
        }
        window.pack()
    }

    fun createResourceTooltip(inventoryImage: InventoryImage) {
        val hoveredTable = createDefaultTooltip(inventoryImage)
        window.add(hoveredTable)

        window.add().row()
        window.add(createLabel(EMPTY_ROW, Color.WHITE)).row()
        val description = inventoryImage.inventoryItem.description.joinToString(System.lineSeparator())
        window.add(createLabel(description, Color.WHITE))
    }

    open fun createSingleTooltip(inventoryImage: InventoryImage) {
        val hoveredTable = createDefaultTooltip(inventoryImage)
        addPossibleDescription(inventoryImage, hoveredTable)
        window.add(hoveredTable)
    }

    private fun createDefaultTooltip(inventoryImage: InventoryImage): Table {
        val hoveredTable = Table()
        hoveredTable.defaults().align(Align.left)

        val totalMerchant = gameData.party.getSumOfSkill(SkillItemId.MERCHANT)
        val descriptionList = inventoryImage.getSingleDescription(totalMerchant).toMutableList()
        removeLeftUnnecessaryAttributes(descriptionList)
        descriptionList.forEach { addToTable(hoveredTable, it, createSingleLabelStyle(it)) }
        return hoveredTable
    }

    private fun createDualTooltip(hoveredImage: InventoryImage, equippedImage: InventoryImage) {
        val hoveredTable = createLeftTooltip(hoveredImage, equippedImage)
        val equippedTable = createRightTooltip(equippedImage, hoveredImage)
        window.add(hoveredTable)
        window.add().spaceRight(HALF_SPACING)
        window.add(equippedTable)
    }

    private fun createLeftTooltip(hoveredImage: InventoryImage, equippedImage: InventoryImage): Table {
        val hoveredTable = Table(window.skin).apply {
            background = Utils.createTooltipRightBorder()
            padRight(HALF_SPACING)
            defaults().align(Align.left)
            add(createLabel(LEFT_TITLE, Color.WHITE)).row()
        }
        val totalMerchant = gameData.party.getSumOfSkill(SkillItemId.MERCHANT)
        val descriptionList = hoveredImage.getDualDescription(equippedImage, totalMerchant).toMutableList()
        removeLeftUnnecessaryAttributes(descriptionList)
        descriptionList.forEach { addToTable(hoveredTable, it, createLeftLabelStyle(it)) }
        addPossibleDescription(hoveredImage, hoveredTable)
        return hoveredTable
    }

    private fun createRightTooltip(equippedImage: InventoryImage, hoveredImage: InventoryImage): Table {
        val equippedTable = Table().apply {
            defaults().align(Align.left)
            add(createLabel(RIGHT_TITLE, Color.LIGHT_GRAY)).row()
        }

        val totalMerchant = gameData.party.getSumOfSkill(SkillItemId.MERCHANT)
        val descriptionList = equippedImage.getDualDescription(hoveredImage, totalMerchant).toMutableList()
        removeRightUnnecessaryAttributes(descriptionList)
        descriptionList.forEach { addToTable(equippedTable, it, createRightLabelStyle(it)) }
        addPossibleDescription(equippedImage, equippedTable)
        return equippedTable
    }

    fun addPossibleDescription(inventoryImage: InventoryImage, table: Table) {
        if (inventoryImage.inventoryItem.description.isNotEmpty()) {
            val description = inventoryImage.inventoryItem.description.joinToString(System.lineSeparator())
            table.add(createLabel(description, Color.VIOLET)).colspan(2)
        }
    }

    fun addToTable(hoveredTable: Table, attribute: InventoryDescription, labelStyle: LabelStyle?) {
        hoveredTable.add(Label(getKey(attribute), labelStyle)).spaceRight(COLUMN_SPACING)
        hoveredTable.add(Label(getValue(attribute), labelStyle)).row()
    }

    open fun removeLeftUnnecessaryAttributes(descriptionList: MutableList<InventoryDescription>) {
        removeBuy(descriptionList)
        removeSell(descriptionList)
    }

    open fun removeRightUnnecessaryAttributes(descriptionList: MutableList<InventoryDescription>) {
        removeBuy(descriptionList)
        removeSell(descriptionList)
    }

    fun removeBuy(descriptionList: MutableList<InventoryDescription>) {
        descriptionList.removeIf { it.key == Constant.DESCRIPTION_KEY_BUY }
        descriptionList.removeIf { it.key == Constant.DESCRIPTION_KEY_BUY_PIECE }
        descriptionList.removeIf { it.key == Constant.DESCRIPTION_KEY_BUY_TOTAL }
    }

    fun removeSell(descriptionList: MutableList<InventoryDescription>) {
        descriptionList.removeIf { it.key == Constant.DESCRIPTION_KEY_SELL }
        descriptionList.removeIf { it.key == Constant.DESCRIPTION_KEY_SELL_PIECE }
        descriptionList.removeIf { it.key == Constant.DESCRIPTION_KEY_SELL_TOTAL }
    }

    fun createSingleLabelStyle(attribute: InventoryDescription): LabelStyle {
        return if (isBuyOrSellValue(attribute)) {
            createLabelStyle(Color.GOLD)
        } else when (attribute.compare) {
            ThreeState.SAME, ThreeState.MORE -> createLabelStyle(Color.WHITE)
            ThreeState.LESS -> createLabelStyle(Color.RED)
        }
    }

    private fun createLeftLabelStyle(attribute: InventoryDescription): LabelStyle {
        return if (isBuyOrSellValue(attribute)) {
            createLabelStyle(Color.GOLD)
        } else when (attribute.compare) {
            ThreeState.SAME -> createLabelStyle(Color.WHITE)
            ThreeState.LESS -> createLabelStyle(ORANGE)
            ThreeState.MORE -> createLabelStyle(Color.LIME)
        }
    }

    private fun createRightLabelStyle(attribute: InventoryDescription): LabelStyle {
        return if (isBuyOrSellValue(attribute)) {
            createLabelStyle(Color.GOLD)
        } else createLabelStyle(Color.WHITE)
    }

    private fun getKey(attribute: InventoryDescription): String {
        return if (attribute.key is SuperEnum) {
            attribute.key.title
        } else {
            attribute.key.toString()
        }
    }

    private fun getValue(attribute: InventoryDescription): String {
        return when {
            attribute.value is SkillItemId -> attribute.value.title
            attribute.key == CalcAttributeId.BASE_HIT -> String.format("%s%%", attribute.value)
            else -> attribute.value.toString()
        }
    }

    private fun createLabel(text: String, color: Color): Label {
        return Label(text, createLabelStyle(color))
    }

    private fun createLabelStyle(color: Color): LabelStyle {
        return LabelStyle(BitmapFont(), color)
    }

    private fun isBuyOrSellValue(attribute: InventoryDescription): Boolean {
        return attribute.key == Constant.DESCRIPTION_KEY_BUY_TOTAL
                || attribute.key == Constant.DESCRIPTION_KEY_SELL_TOTAL
                || attribute.key == Constant.DESCRIPTION_KEY_BUY_PIECE
                || attribute.key == Constant.DESCRIPTION_KEY_SELL_PIECE
                || attribute.key == Constant.DESCRIPTION_KEY_BUY
                || attribute.key == Constant.DESCRIPTION_KEY_SELL
    }

}
